doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1746,33 +1784,77 @@ Controlling the proof flow
+   Tries to prove the current goal by finding a contradiction.
 
-   .. exn:: No such assumption.
-      :undocumented:
+   If :n:`@one_term_with_bindings` is not provided, the tactic first does an
+   :tacn:`intros`.  The tactic then proves the goal if the updated context has
+   a pair of hypotheses where one is the negation of the other (e.g. :n:`P` and
+   not :n:`~P`) or if there is a hypothesis with the type :n:`False` (or another
+   empty inductive type).
+
+   If :n:`@one_term_with_bindings` is provided: The type of
+   :n:`@one_term_with_bindings` must be a negation, such as :n:`~P`,
+   or :n:`False` (or another empty inductive type).
+   If the type is a negation and :n:`P` is a hypothesis in the context,
+   the goal is proven.  If the type is a negation and :n:`P` is not in
+   the context, the goal is replaced with :n:`P`.  If the type is :n:`False`,
+   the goal is proven.  Otherwise the tactic fails.  (In particular, note
+   that the converse doesn't work, e.g. if the type is :n:`P`, the tactic
+   won't find :n:`~P`.)
+
+   Use the :tacn:`discriminate` tactic to prove the current goal when there
+   is a hypothesis with an impossible structural equality such as
+   :n:`0 = 1`.
 
-.. tacv:: contradiction @ident
+.. example:: :tacn:`contradiction` tactic
+
+   Simple examples.  To see more detail, add `intros` after each `Goal`.
+
+   .. coqtop:: reset in
+
+      Inductive F :=. (* Another empty inductive type *)
+
+      Goal F -> False.
+      contradiction.
+      Qed.
+
+      Goal forall (A : Prop), A -> ~A -> False.
+      contradiction.
+      Qed.
+
+      Goal forall (A : Type) (x : A), ~(x = x) -> False.
+      contradiction.
+      Qed.
+
+   Apply a fact from the standard library:
+
+   .. coqtop:: in
+
+      Require Import Arith.
+
+      Goal forall (A : Prop), 0 < 0 -> A.
+
+   .. coqtop:: all
 
-   The proof of False is searched in the hypothesis named :n:`@ident`.
+      intros. (* the intros is needed because 0 < 0 is not a negation
+                 while the type of (Nat.lt_ irrefl 0) is a negation (~ 0 < 0) *)
+      contradiction (Nat.lt_irrefl 0).
+      Qed.
 
 .. tacn:: contradict @ident
-   :name: contradict
 
-   This tactic allows manipulating negated hypothesis and goals. The name
-   :n:`@ident` should correspond to a hypothesis. With :n:`contradict H`, the
-   current goal and context is transformed in the following way:
+   Transforms the specified hypothesis `:n:`@ident` and the goal in order to
+   prove that the hypothesis is false. For :n:`contradict H`, the
+   current goal and context are transformed as shown.  (For brevity,
+   `⊢` is used to separate hypotheses from the goal; it is equivalent to the
+   dividing line shown in a context.):
 
-   + H:¬A ⊢ B becomes ⊢ A
-   + H:¬A ⊢ ¬B becomes H: B ⊢ A
-   + H: A ⊢ B becomes ⊢ ¬A
-   + H: A ⊢ ¬B becomes H: B ⊢ ¬A
+   + H: ~A ⊢ B becomes ⊢ A
+   + H: ~A ⊢ ~B becomes H: B ⊢ A
+   + H: A ⊢ B becomes ⊢ ~A
+   + H: A ⊢ ~B becomes H: B ⊢ ~A
```

Comment:
I don't think that the removal of the Unicode notation for negation is making the reading any better, especially given that none of these are in a verbatim context.

Reply:
It's more consistent.  This is the only place in the doc that uses ¬.

Reply:
But isn't it also one of the few places that uses ⊢? I would think this symbol is otherwise probably only used in the typing rules section, and in this section the negation would certainly be represented as ¬ and not as ~.
If you really want to keep the ~ (and I really think this does not help readability), then put some quotes so that this is rendered as code.

Reply:
Turnstile is used a lot in the more mathematical sections and very little elsewhere.  `¬` is used in the doc only once in `coqdoc.rst`.

![image](https://user-images.githubusercontent.com/1253341/209216130-97c5c6a2-a8dd-4ac3-b81b-d0a3e8ac35a4.png)

> put some quotes so that this is rendered as code

Like this?

![image](https://user-images.githubusercontent.com/1253341/209218398-a35d0a2c-d9ed-4f83-9a7b-bff67d9c2d6b.png)

Reply:
Yes